src/compiler/transformers/es2018.ts

```diff
@@ -59,6 +59,7 @@ namespace ts {
         let exportedVariableStatement = false;
         let enabledSubstitutions: ESNextSubstitutionFlags;
         let enclosingFunctionFlags: FunctionFlags;
+        let parameterHasPrecedingObjectRestOrSpread: Set<ParameterDeclaration> | undefined;

```

Comment:
I don’t understand why this is shared state when every function that uses it carefully sets and resets it so as to pretend it’s local state. Can’t `visitConstructorDeclaration` etc. simply create this Set locally and pass it to `visitParameter`?

Reply:
We don't pass state through `visitParameterList`, so there's no way to pass it to `visitParameter` without duplicating the logic of `visitParameterList`.

Reply:
The only other alternative is to use an arrow function that creates a closure, and that would affect performance.

Reply:
Does this even work correctly when there are parameters nested within parameters?

```ts
function f({ ...a, x }, b = (c, d) => {}) {}
```

Won’t `c` and `d` be added to the Set containing `b`? It looks to me like each function also needs to set it to `undefined` after saving it, which is even more bookkeeping ☹️

Reply:
While I don't disagree, it's a pattern we employ throughout the transformers. That said, you are correct that the child parameters are added to the set incorrectly. I've amended the logic to ensure the set is scoped to each function/method.